File: sections/100_grandmaster_finale/problems/14_rooted_broadcast_choice/solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct Edge {
    int from, to;
    long long cost;
};

long long directed_mst(int vertices, int root, vector<Edge> edges) {
    const long long INF = (1LL << 62);
    long long answer = 0;
    while (true) {
        vector<long long> incoming(vertices, INF);
        vector<int> parent(vertices, -1);
        for (const auto &edge : edges) {
            if (edge.from != edge.to && edge.cost < incoming[edge.to]) {
                incoming[edge.to] = edge.cost;
                parent[edge.to] = edge.from;
            }
        }
        incoming[root] = 0;
        for (int v = 0; v < vertices; ++v) {
            if (incoming[v] == INF) return -1;
            answer += incoming[v];
        }

        int cycles = 0;
        vector<int> component(vertices, -1), seen(vertices, -1);
        for (int start = 0; start < vertices; ++start) {
            int v = start;
            while (seen[v] != start && component[v] == -1 && v != root) {
                seen[v] = start;
                v = parent[v];
            }
            if (v != root && component[v] == -1) {
                for (int u = parent[v]; u != v; u = parent[u]) {
                    component[u] = cycles;
                }
                component[v] = cycles++;
            }
        }
        if (cycles == 0) return answer;

        for (int v = 0; v < vertices; ++v) {
            if (component[v] == -1) component[v] = cycles++;
        }
        vector<Edge> contracted;
        contracted.reserve(edges.size());
        for (const auto &edge : edges) {
            int from = component[edge.from];
            int to = component[edge.to];
            long long cost = edge.cost;
            if (from != to) cost -= incoming[edge.to];
            contracted.push_back({from, to, cost});
        }
        root = component[root];
        vertices = cycles;
        edges.swap(contracted);
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;
    vector<long long> activation(n);
    for (long long &cost : activation) cin >> cost;

    const long long PENALTY = (long long)(n + 1) * 1'000'000'000LL + 1;
    vector<Edge> edges;
    edges.reserve(m + n);
    for (int i = 0; i < m; ++i) {
        int u, v;
        long long cost;
        cin >> u >> v >> cost;
        edges.push_back({u - 1, v - 1, cost});
    }
    int super_root = n;
    for (int v = 0; v < n; ++v) {
        edges.push_back({super_root, v, PENALTY + activation[v]});
    }

    long long encoded = directed_mst(n + 1, super_root, edges);
    if (encoded < 0 || encoded / PENALTY != 1) cout << -1 << '\n';
    else cout << encoded % PENALTY << '\n';
}

File: sections/100_grandmaster_finale/problems/14_rooted_broadcast_choice/solution.py
import sys


def directed_mst(vertices: int, root: int,
                 edges: list[tuple[int, int, int]]) -> int | None:
    answer = 0
    infinity = 10**40
    while True:
        incoming = [infinity] * vertices
        parent = [-1] * vertices
        for u, v, cost in edges:
            if u != v and cost < incoming[v]:
                incoming[v] = cost
                parent[v] = u
        incoming[root] = 0
        if any(cost == infinity for cost in incoming):
            return None
        answer += sum(incoming)

        component = [-1] * vertices
        seen = [-1] * vertices
        cycles = 0
        for start in range(vertices):
            v = start
            while seen[v] != start and component[v] == -1 and v != root:
                seen[v] = start
                v = parent[v]
            if v != root and component[v] == -1:
                u = parent[v]
                while u != v:
                    component[u] = cycles
                    u = parent[u]
                component[v] = cycles
                cycles += 1
        if cycles == 0:
            return answer

        for v in range(vertices):
            if component[v] == -1:
                component[v] = cycles
                cycles += 1
        contracted = []
        for u, v, cost in edges:
            new_u, new_v = component[u], component[v]
            if new_u != new_v:
                cost -= incoming[v]
            contracted.append((new_u, new_v, cost))
        root = component[root]
        vertices = cycles
        edges = contracted


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n, m = data[:2]
    activation = data[2:2 + n]
    edges = []
    index = 2 + n
    for _ in range(m):
        u, v, cost = data[index:index + 3]
        index += 3
        edges.append((u - 1, v - 1, cost))

    penalty = (n + 1) * 10**9 + 1
    super_root = n
    edges.extend((super_root, v, penalty + activation[v]) for v in range(n))
    encoded = directed_mst(n + 1, super_root, edges)
    if encoded is None or encoded // penalty != 1:
        print(-1)
    else:
        print(encoded % penalty)


if __name__ == "__main__":
    main()
